### ISSUE TITLE
Fix `pack` issues introduced in zip64 extension in Archive::Zip 1.66

### ISSUE THREAD

**Issue by farblos** [redhotpenguin/perl-Archive-Zip#56]
The zip64 extension in Archive::Zip, introduced with version 1.66, uses some "modern" `pack` codes that let `make test` fail.  This issue tracks the problems:

- `Q` is not available on 32-bit Perls

- modifiers `<` and `>` are not available on Perl 5.8

**Comment by farblos** [comment]
@redhotpenguin, have you already done anything about these? If not, I could take over. My proposal:

1. Throw errors when zip64 is needed on any of the Perls mentioned above.

2. Skip all zip64-related tests on any of the Perls mentioned above.

Personally, I don't think that it would be worth to replace the current `pack` calls by ugly, but backward-compatible constructs.

**Comment by pmqs** [comment]
As things stand `Archive::Zip` is _completely_ broken on all 32-bit builds and all 5.8.* releases. There are a number of options

1. Don't include support for Zip64 on 32-bit/5.8 _at all_ 
This means we have a module that works on 32/64-bit builds and all supported Perl versions. Support for Zip64 can be added later if someone has the tuits.


1. Add Zip64 support for either 5.8 and/or 32-bit builds.
I did it in `IO::Compress::Zip`, but it was a real pain to write. The code to handle the 64-bit manipulation all lives in a single class, called `U64`.  It should be self-contained enough to reuse, if anyone is interested.

Unless someone has the tuits to address the 32-bit/5.8 issues right now, I say we go with your proposal @farblos

**Comment by redhotpenguin** [comment]
Please feel free to take over

**Comment by farblos** [comment]
> Please feel free to take over

Will do.

@redhotpenguin, could you please merge PRs #53 and #57 soon? I'm still getting used to git/hub and I'd feel better if I could see these changes in the code base already before starting off with my fixes. Thx!

**Comment by redhotpenguin** [comment]
Done

**Comment by farblos** [comment]
@pmqs: On my dev system "Debian GNU/Linux 9 (stretch)" with package `unzip` as of 6.0-21+deb9u2, the new bzip2 test triggers some "bomb detection" error:

```
#   Failed test 'testZip ok after t/data/bzip.zip to /home/jschmidt/android/tmp/testout-KOEvs.zip'
#   at t/26_bzp2.t line 33.
#          got: '1024'
#     expected: '0'
ziptest said: Archive:  /home/jschmidt/android/tmp/testout-KOEvs.zip
    testing: hello.txt                OK
error: not enough memory for bomb detection
```

Actually, the unzip fails already on the input test file:

```
[perl-Archive-Zip]$ unzip -t t/data/bzip.zip 
Archive:  t/data/bzip.zip
    testing: hello.txt                OK
error: not enough memory for bomb detection
At least one error was detected in t/data/bzip.zip.
```

Do you think you can cobble up a bzip2 file that would not trigger the "bomb detection"?  BTW, https://github.com/madler/unzip seems to cover exactly that new logic.

I can test any bzip2 files if you cannot reproduce the error ...

**Comment by farblos** [comment]
Just to clarify: I've tested *with* your latest PR #57 in place. The problem is not the PR being present or not, the problem is that `unzip -t` is not good for testing the resulting zip file on my system.

**Comment by pmqs** [comment]
> Do you think you can cobble up a bzip2 file that would not trigger the "bomb detection"?

Yes, See #58

**Comment by farblos** [comment]
Thx.  That was quick.  Not so quickly working on the main issue ...

**Comment by farblos** [comment]
I'm done, @redhotpenguin.

I've created PR #59 (for fixing the pack issue) and PR #60 (to provide a draft for updating Changes).

Please merge and push 1.67 only if Travis comes through green.

Version 1.67 would close issues #51, #54, #65.

**Comment by farblos** [comment]
> Version 1.67 would close issues #51, #54, #65.

Version 1.67 would close issues #51, #54, ***#56***.

**Comment by farblos** [comment]
And, of course, it fixes issue #18!

**Comment by redhotpenguin** [comment]
1.67 is enroute to CPAN

**Comment by farblos** [comment]
Thanks.  Will wait for the cpantesters results pouring in and close this issue if things get as green as pre-1.66.

@redhotpinguin, could you please close issue #18? And while you're at it, you could as well close with reference to version 1.67:
- [RT #70856 - Add Zip64 support](https://rt.cpan.org/Public/Bug/Display.html?id=70856)
- [RT #78253 - 4GB limitation of zip file or data](https://rt.cpan.org/Public/Bug/Display.html?id=78253)
- [RT #116471 - Problem with Archive-Zip and very large files](https://rt.cpan.org/Public/Bug/Display.html?id=116471)

That would feel, um, rewarding.

**Comment by pmqs** [comment]
@farblos so far the majority of cpantestsers failure are for `26_bzip2.t`. That should be addressed by PR https://github.com/redhotpenguin/perl-Archive-Zip/pull/63. 

The only failure I see so far that is Zip64-related is http://www.cpantesters.org/cpan/report/6fdfbb90-e994-11e9-a367-d8d5ba5512dc. It is failing quite a few tests, including this one

```
#   Failed test at t/21_zip64.t line 134.
#          got: '512'
#     expected: '0'

#   Failed test at t/21_zip64.t line 159.
#          got: '20736'
#     expected: '0'
# Looks like you failed 2 tests of 80.
t/21_zip64.t .................. 
Dubious, test returned 2 (wstat 512, 0x200)
```

**Comment by redhotpenguin** [comment]
Rt tickets updated. Felt rewarding :)

**Comment by pmqs** [comment]
> Rt tickets updated. Felt rewarding :)

@redhotpenguin that reminds me - the [metacpan link](https://metacpan.org/pod/Archive::Zip) for this module points to [rt.cpan.org](https://rt.cpan.org/Dist/Display.html?Status=Active;Queue=Archive-Zip) for issues, rather than here. Do you want that to remain the official home for bug reports?

I see that  https://github.com/redhotpenguin/perl-Archive-Zip/issues/24 is asking the same question.

**Comment by redhotpenguin** [comment]
I think it makes sense to move it to Github for now.
